pkg/services/ngalert/state/cache.go

```diff
@@ -34,23 +34,54 @@ func newCache() *cache {
 }
 
 func (c *cache) getOrCreate(ctx context.Context, log log.Logger, alertRule *ngModels.AlertRule, result eval.Result, extraLabels data.Labels, externalURL *url.URL) *State {
+	// Calculation of state ID involves label and annotation expansion, which may be resource intensive operations, and doing it in the context guarded by mtxStates may create a lot of contention.
+	// Instead of just calculating ID we create an entire state - a candidate. If rule states already hold a state with this ID, this candidate will be discarded and the existing one will be returned.
+	// Otherwise, this candidate will be added to the rule states and returned.
+	stateCandidate := calculateState(ctx, log, alertRule, result, extraLabels, externalURL)
+
 	c.mtxStates.Lock()
 	defer c.mtxStates.Unlock()
+
 	var orgStates map[string]*ruleStates
 	var ok bool
-	if orgStates, ok = c.states[alertRule.OrgID]; !ok {
+	if orgStates, ok = c.states[stateCandidate.OrgID]; !ok {
 		orgStates = make(map[string]*ruleStates)
-		c.states[alertRule.OrgID] = orgStates
+		c.states[stateCandidate.OrgID] = orgStates
 	}
 	var states *ruleStates
-	if states, ok = orgStates[alertRule.UID]; !ok {
+	if states, ok = orgStates[stateCandidate.AlertRuleUID]; !ok {
 		states = &ruleStates{states: make(map[string]*State)}
-		c.states[alertRule.OrgID][alertRule.UID] = states
+		c.states[stateCandidate.OrgID][stateCandidate.AlertRuleUID] = states
+	}
+	return states.getOrAdd(stateCandidate)
+}
+
+func (rs *ruleStates) getOrAdd(stateCandidate State) *State {
+	state, ok := rs.states[stateCandidate.CacheID]
+	// Check if the state with this ID already exists.
+	if !ok {
+		rs.states[stateCandidate.CacheID] = &stateCandidate
+		return &stateCandidate
+	}
+
+	// Annotations can change over time, however we also want to maintain
+	// certain annotations across evaluations
+	for k, v := range state.Annotations {
+		if _, ok := ngModels.InternalAnnotationNameSet[k]; ok {
+			// If the annotation is not present then it should be copied from the
+			// previous state to the next state
+			if _, ok := stateCandidate.Annotations[k]; !ok {
```

Comment:
Do you have an example of some annotations that shouldn't change between evaluations?

Reply:
This was introduced https://github.com/grafana/grafana/pull/49080, I just moved code to a function as is. IIRC, it was introduced to persist image annotation.